[SCHEMA] Add metadata term files

tsalo opened the pull request:
References #699.

This is a draft that might be useful for determining how to structure metadata YAML files for the schema. 

This approach does not incorporate any interactions between metadata terms, or requirement levels, which are both dependent on the suffix or data type to which the metadata term is being applied. This fits well with #603.

Changes proposed:

- Take the metadata JSON files from [the bids-validator](https://github.com/bids-standard/bids-validator/tree/master/bids-validator/validators/json/schemas) and convert them to YAML format.
- Add "description" and "key_name" fields to the metadata YAML files.
- Add functions to render the metadata information in tables within the schema. These functions require the specification text to include the requirement levels.
- Create files for a number of metadata terms that aren't present in the validator, but _are_ present in specification tables.

TODO:

- [x] Different units dependent on different data types?
    - Yes
- [x] Include "unit" for string/integer data types?
    - No
- [x] Limits, like minimum and maximum values
    - Taken from validator
- [x] Should we retain references to other info (e.g., lists of valid coordinate spaces for iEEG, EEG, and MEG)?
    - Currently, I have the references still in the files, with the referenced info in "private" yaml files (i.e., files with leading underscore).
- [x] Links that have to be broken up aren't looking great.
    - This is because the flattening step adds a space between separate lines.
- [x] Would be great to replace relative link paths with absolute ones.
- [X] Type determination for objects.
    - `additionalProperties` defines type.
    - `properties` should create additional rows. How should individual sub-object-level requirement levels be annotated then?? Take `Genetics.Dataset`, `Genetics.Database`, and `Genetics.Descriptors` as an example.
        - SOLUTION: Instead of trying to expand nested dictionaries into new rows, I've replaced the internal definitions with `$ref`s and added new files for the properties that can be referenced on their own. So now Genetics' `properties` fields is a dict with three `$ref`s and there are separate files for `Genetics.Dataset.yaml` etc.
- [x] Make type determination function recursive to deal with nested objects
    - See `SoftwareFilters` as an example.
- [x] Document new structure once #740 is merged.
- [x] Add "relative" and "absolute"(?) path formats for strings?
    - I decided to go with `dataset_relative` and `participant_relative` for now.
- [x] Add "unit" format for strings?
    - With PET now merged in, we have a bunch of new metadata fields that just define units of other fields instead of committing to a standard like we have for most other fields with units. Maybe we could define a "unit" format that will let the validator check for valid units?
- [ ] Identify and incorporate section-specific definition modifications.

tsalo commented on src/schema/metadata/PostLabelingDelay.yaml:
```diff
@@ -0,0 +1,18 @@
+  Based on DICOM Tags 0018,9079 `Inversion Times` and 0018,0082
+  `InversionTime`.
+data_type:
+  - number
```
```suggestion
  - number:
    - minimum: 0
```

What about this kind of notation for limits?

tsalo commented on src/schema/metadata/PostLabelingDelay.yaml:
```diff
@@ -0,0 +1,18 @@
+  `InversionTime`.
+data_type:
+  - number
+  - array of numbers
```
```suggestion
  - array:
      number:
        minimum: 0
```

Maybe this for nesting rules for the data types within the array?

effigies commented:
Is there an advantage to a single file per term? Grouping terms feels more natural to me.

tsalo commented:
I don't want to group terms into multiple files, since relationships between terms (including their categories) probably vary across the specification, along with any logic that needs to be applied. I figured one file per term would be easier to search and edit than one file for all terms.

tsalo commented:
Based on the current code, the following macro call in the specification produces the table at the bottom. Some are missing descriptions right now and others are just missing completely, but it's still pretty cool!

```Text
{{ MACROS___make_metadata_table(
   {
      "Manufacturer": "RECOMMENDED",
      "ManufacturersModelName": "RECOMMENDED",
      "DeviceSerialNumber": "RECOMMENDED",
      "StationName": "RECOMMENDED",
      "SoftwareVersions": "RECOMMENDED",
      "HardcopyDeviceSoftwareVersion": "DEPRECATED",
      "MagneticFieldStrength": "RECOMMENDED, but REQUIRED for Arterial Spin Labeling",
      "ReceiveCoilName": "RECOMMENDED",
      "ReceiveCoilActiveElements": "RECOMMENDED",
      "GradientSetType": "RECOMMENDED",
      "MRTransmitCoilSequence": "RECOMMENDED",
      "MatrixCoilMode": "RECOMMENDED",
      "CoilCombinationMethod": "RECOMMENDED",
   }
) }}
```

![image](https://user-images.githubusercontent.com/8228902/111542146-ba1fb500-8747-11eb-9078-02fc8a4d68dc.png)

tsalo commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -10,6 +10,24 @@ by Ben Inglis:
 
 ### Scanner Hardware
 
+{{ MACROS___make_metadata_table(
+   {
+      "Manufacturer": "RECOMMENDED",
+      "ManufacturersModelName": "RECOMMENDED",
+      "DeviceSerialNumber": "RECOMMENDED",
+      "StationName": "RECOMMENDED",
+      "SoftwareVersions": "RECOMMENDED",
+      "HardcopyDeviceSoftwareVersion": "DEPRECATED",
+      "MagneticFieldStrength": "RECOMMENDED, but REQUIRED for Arterial Spin Labeling",
+      "ReceiveCoilName": "RECOMMENDED",
+      "ReceiveCoilActiveElements": "RECOMMENDED",
+      "GradientSetType": "RECOMMENDED",
+      "MRTransmitCoilSequence": "RECOMMENDED",
+      "MatrixCoilMode": "RECOMMENDED",
+      "CoilCombinationMethod": "RECOMMENDED",
+   }
+) }}
```
Until we add conditional logic to the schema, this would be how we define requirement levels in the specification.

tsalo commented on src/schema/metadata/AnatomicalLandmarkCoordinateSystem.yaml:
```diff
@@ -0,0 +1,13 @@
+  See Appendix VIII for a list of restricted keywords for coordinate systems.
+  If `Other`, provide definition of the coordinate system in
+  `AnatomicalLandmarkCoordinateSystemDescription`.
+anyOf:
+- $ref: common_definitions.json#/definitions/MEGCoordSys
+- $ref: common_definitions.json#/definitions/EEGCoordSys
+- $ref: common_definitions.json#/definitions/StandardTemplateCoordSys
+- $ref: common_definitions.json#/definitions/StandardTemplateDeprecatedCoordSys
```
I'm not so sure about this one. Do we want to include references in the schema? If so, is this the best way to do it?

tsalo replied:
Alternatively, we could make sure that all referenced definitions are in "private" files (e.g., `_ StandardTemplateCoordSys.yaml`), since those definitions shouldn't be useful on their own.

tsalo commented:
I can keep trucking along, adding terms, but I'd like to get input from interested folks.

@satra @dbkeator would this format be useable for NIDM-Terms?

@yarikoptic I know this isn't exactly how you wanted it (e.g., files are grouped in subfolders by type), but WDYT? The actual interactions between terms will still need to be defined elsewhere in the schema.

@rwblair could the validator use files like this? They're basically a translation of the validator JSON files to individual YAML files, with added descriptions and units, so I assume it wouldn't require too much effort to support them.

yarikoptic commented on src/02-common-principles.md:
```diff
@@ -498,6 +498,16 @@ Note that if a field name included in the data dictionary matches a column name
+      "TermURL": "RECOMMENDED",
+   }
+) }}
+
```
table seems to be (rendered) empty, shortcut link https://bids-specification--762.org.readthedocs.build/en/762/02-common-principles.html#tabular-files

yarikoptic commented on src/03-modality-agnostic-files.md:
```diff
@@ -14,6 +14,23 @@ Templates:
 
+{{ MACROS___make_metadata_table(
+   {
+      "Name": "REQUIRED",
```
I don't see name in rendered version: https://bids-specification--762.org.readthedocs.build/en/762/03-modality-agnostic-files.html#dataset_descriptionjson

yarikoptic commented on src/03-modality-agnostic-files.md:
```diff
@@ -69,6 +86,13 @@ In addition to the keys for raw BIDS datasets,
+{{ MACROS___make_metadata_table(
+   {
+      "GeneratedBy": "REQUIRED",
+      "SourceDatasets": "RECOMMENDED",
```
SourceDatasets is missing: https://bids-specification--762.org.readthedocs.build/en/762/03-modality-agnostic-files.html#derived-dataset-and-pipeline-description

I guess script should be adjusted to error out if some field is not "handled"?

tsalo replied:
Ultimately yes, but since I'm still in the process of converting all of the metadata terms, I don't want everything to error out just yet. Right now it just filters out any missing terms and renders the available ones.

yarikoptic commented on src/03-modality-agnostic-files.md:
```diff
@@ -14,6 +14,23 @@ Templates:
+      "EthicsApprovals": "OPTIONAL",
+      "ReferencesAndLinks": "OPTIONAL",
+      "DatasetDOI": "OPTIONAL",
+   }
```
having such specifications in the document text would be a big step back from toward the target of having a "machine readable" schema.

Why not to add `metadata` key to the corresponding records, e.g. in this case of https://github.com/bids-standard/bids-specification/blob/master/src/schema/top_level_files.yaml#L14 where similar to `entities` of https://github.com/bids-standard/bids-specification/tree/master/src/schema/datatypes to list the 'requirements' for each one of those? and then in the macros just point to what metadata of to render here (e.g. of `top_level_files.dataset_description` here)?  or you can immediately see that this would not work?

tsalo replied:
> having such specifications in the document text would be a big step back from toward the target of having a "machine readable" schema.

I see it as an incremental move forward. It definitely doesn't solve a lot of the problems (e.g., requirement levels, conditional relationships between metadata fields), but to be fair those are the _hard_ problems so I don't feel too bad about it.

> Why not to add metadata key to the corresponding records, e.g. in this case of https://github.com/bids-standard/bids-specification/blob/master/src/schema/top_level_files.yaml#L14 where similar to entities of https://github.com/bids-standard/bids-specification/tree/master/src/schema/datatypes to list the 'requirements' for each one of those? and then in the macros just point to what metadata of to render here (e.g. of top_level_files.dataset_description here)? or you can immediately see that this would not work?

I'll have to try it out, but that sounds like a good plan for the PR _after_ this one. I figure that, since we want to separate the core definitions from the rules of the schema, we can tackle those two pieces in separate stages.

tsalo replied:
I think the best next step for the schema (after this) is to try to identify the types of rules we can expect from the schema, along with any really weird edge cases. That's what I've tried to do in #620, although I worry that my current list isn't exhaustive. Without a list of necessary rules, I'm not sure if we can define them in YAML files in a way that will satisfy everyone.

yarikoptic replied:
Sure could and should be a multistage process. I am ok with your proposal. I just thought that  it might be nice to just put rules identification for later pr, thus keep description free form, while already allowing for machine readable specification of possible requirement levels and keeping everything under src/schema. But such a move could indeed be just yet another follow up PR to not block this one

yarikoptic commented on src/04-modality-specific-files/04-intracranial-electroencephalography.md:
```diff
@@ -362,12 +423,28 @@ also be specified.
+   {
+      "iEEGCoordinateSystem": "REQUIRED",
+      "iEEGCoordinateUnits": "REQUIRED",
+      "iEEGCoordinateSystemDescription": 'RECOMMENDED, but REQUIRED if `iEEGCoordinateSystem` is `"Other"`',
```
oh, that is an interesting case...  in light of my eager desire for concentrating this information in machine readable form, I can suggest for such cases to get to temp middle point  where requirement could be "spelled out" in human form for now with the last one listed as "catch all", e.g. 

```yaml
metadata:
 iEEGCoordinateUnits: required
 iEEGCoordinateUnits: required
 iEEGCoordinateSystemDescription:
  required: if `iEEGCoordinateSystem` is `"Other"`
  recommended:
 iEEGCoordinateProcessingDescription: recommended
```

yarikoptic reviewed (changes requested):
> @yarikoptic I know this isn't exactly how you wanted it (e.g., files are grouped in subfolders by type), but WDYT? The actual interactions between terms will still need to be defined elsewhere in the schema.

GREAT STUFF HERE, thank you @tsalo for all the work on this!  I blissfully forgotten (and decided to not encumber my review with prior opinion by looking it up) what I wanted.  But I immediately saw what I would have really preferred to avoid here -- "leaking" schema into the document.   I have dropped some question/idea(s) in other comments on how possibly may be those metadata fields "requirements" could be centralized within schema.  WDYT?

rwblair commented:
> @rwblair could the validator use files like this? They're basically a translation of the validator JSON files to individual YAML files, with added descriptions and units, so I assume it wouldn't require too much effort to support them.

The validator certainly could build its json schema files from this. Is there any where encoded in yaml which terms are acceptable in which json files?

tsalo commented:
>Is there any where encoded in yaml which terms are acceptable in which json files?

Not yet. I think I'll need to translate over the context files from the validator, with some minor additions.

sappelhoff commented:
I think the URL the linkchecker flags has moved to `https://www.hedtags.org/docs/HEDTaggingStrategyGuide.pdf`, we should probably fix that in a PR that will be merged soon, or make a dedicated PR

tsalo commented:
>The validator certainly could build its json schema files from this. Is there any where encoded in yaml which terms are acceptable in which json files?

@rwblair, I'm sorry, it looks like I misunderstood your question. Right now the _rules_ applied to the metadata terms (e.g., which terms apply to each suffix or datatype) are not implemented, and I wasn't planning to put that in this PR. I figured that the validator could retain those rules for now, but without the information that is in the YAML files I'm proposing.

I'm hopeful that someone will be able to tackle the rules (and thus #699) in a future PR.

tsalo commented:
Okay, I think I'm happy with this, so I'm marking it as ready for review.

tsalo commented on src/schema/metadata/PowerLineFrequency.yaml:
```diff
@@ -0,0 +1,11 @@
+  - type: number
+    exclusiveMinimum: 0
+    unit: Hz
+  - type: string
+    pattern: ^n/a$
```
Could we replace the `^n/a$` pattern with a one-item `enum`?

sappelhoff replied:
yes, technically that would be equivalent except for the message that a user gets upon validating that schema. (I.e., "foo must be pattern X" versus "foo must be one of X" or the like). But that small difference would be no problem IMHO.

tsalo replied:
The current issue is that the tables render these kinds as `object or string` instead of `object or "n/a"`. I thought maybe checking against an enum list and rendering those values directly if the list is short would be easier than figuring something out for pattern strings, but maybe it'll be easier if I figure out what the characters in the pattern represent, and what patterns are possible?

sappelhoff replied:
> maybe it'll be easier if I figure out what the characters in the pattern represent, and what patterns are possible?

There are some libraries that allow you to generate strings from a given regular expression, see: https://github.com/asciimoo/exrex

So perhaps it wouldn't even be too difficult. And coincidentally, it may provide us with a nice double check of our regexps :-)

tsalo commented on src/schema/metadata/BackgroundSuppressionNumberPulses.yaml:
```diff
@@ -0,0 +1,8 @@
+  Note that this excludes any effect of background suppression pulses applied
+  before the labeling.
+type: integer
+minimum: 0
```
Is `minimum` equivalent to `inclusiveMinimum`?

sappelhoff replied:
> Is minimum equivalent to inclusiveMinimum?

conceptually yes, but according to the JSON schema spec, there is no dedicated keyword for `inclusiveMinimum`: https://json-schema.org/understanding-json-schema/reference/numeric.html?highlight=minimum#range



        x ≥ minimum
        x > exclusiveMinimum
        x ≤ maximum
        x < exclusiveMaximum

sappelhoff replied:
related: https://github.com/bids-standard/bids-validator/pull/1261

tsalo replied:
Awesome! I'll replace `Minimum` with `minimum` and `inclusiveMaximum` with `maximum`. Thanks!

tsalo commented on src/schema/metadata/BolusCutOffDelayTime.yaml:
```diff
@@ -0,0 +1,20 @@
+  Based on DICOM Tag 0018,925F `ASL Bolus Cut-off Delay Time`.
+anyOf:
+  - type: number
+    Minimum: 0
```
I've noticed there's both `minimum` and `Minimum`.

sappelhoff replied:
That's weird, I am pretty sure that in JSON schema, the keys are case sensitive, and that `minimum` is a key and `Minimum` is not. I am just surprised that we did not run into errors so far.

will fix in https://github.com/bids-standard/bids-validator/pull/1261

tsalo commented on src/schema/metadata/IntendedFor.yaml:
```diff
@@ -0,0 +1,14 @@
+  This field is OPTIONAL, and in case the fieldmaps do not correspond to any
+  particular scans, it does not have to be filled.
+anyOf:
+  - type: string
+    minLength: 1
```
What if we added a `format: relative_path` option?

tsalo replied:
I've added `dataset_relative`, `participant_relative`, and `unit` formats. They may not be used yet, but they could ultimately be useful, I think.

tsalo commented on src/schema/metadata/MEGCoordinateSystem.yaml:
```diff
@@ -0,0 +1,13 @@
+  If `"Other"`, provide definition of the coordinate system in
+  `MEGCoordinateSystemDescription`.
+anyOf:
+  - $ref: _MEGCoordSys.yaml
+  - $ref: _EEGCoordSys.yaml
+  - $ref: _StandardTemplateCoordSys.yaml
+  - $ref: _StandardTemplateDeprecatedCoordSys.yaml
```
This currently renders as "string or string or string or string". Not sure how best to clean it up.

rwblair replied:
I tried rendering it with the enum typed fields joined with " or ", unfortunately the resulting list seems unwieldy to put directly into the table. This also wouldn't make sense for other fields like GeneticLevel where the enumerations are listed directly in the description.

Is there any current situation where removing duplicates of  what `_resolve_metadata_type` returns when generating the string for anyOf would cause a problem?

tsalo replied:
Currently it's just a single single that's appended to, so it's difficult to identify and remove duplicates. I could reorganize it to a list of lists or something that would be easier to identify and remove duplicates though.

tsalo commented on src/schema/metadata/Genetics.yaml:
```diff
@@ -0,0 +1,12 @@
+  An object containing information about the genetics descriptor.
+type: object
+properties:
+  Database:
+    $ref: Genetics.Database.yaml
+  Dataset:
+    $ref: Genetics.Dataset.yaml
+  Descriptors:
+    $ref: Genetics.Descriptors.yaml
```
What does everyone think of this instead of defining those fields in here directly?

sappelhoff replied:
looks clean to me

sappelhoff reviewed (commented):
How can we most effectively review these 234 files? Do you have a "guide" @tsalo ? Where is another set of eyes most useful?

tsalo commented:
I think the best place to start is the inline comments/questions I added. I also added a lot of information to `src/schema/README.md` that I _hope_ is accurate, and I'd love input on it. Other than that, I guess the best things to do would be to check the generated tables, which are rendered alongside the original tables for easy comparison.

tsalo commented on src/schema/metadata/Columns.yaml:
```diff
@@ -0,0 +1,8 @@
+type: array
+items:
+  type: string
+  minLength: 1
```
Could we just assume that strings have a minimum length of 1? If the field is provided with a string, then I think it should have at least one character to be valid, from a BIDS standpoint. Missing data should be provided with `"n/a"`, so there's not really a case for `""` appearing in jsons.

sappelhoff replied:
agreed! And then users can still abuse it by providing `" "`, but it might be sufficient to make them think twice :-)

tsalo replied:
I think the same can apply to `minItems` when the type is `array`, so I'll remove all cases of `minLength: 1` and `minItems: 1` from the YAML files.

tsalo commented on src/schema/metadata/DatasetDOI.yaml:
```diff
@@ -0,0 +1,10 @@
+  DOIs SHOULD be expressed as a valid
+  [URI](/02-common-principles.html#uniform-resource-indicator);
+  bare DOIs such as `10.0.2.3/dfjj.10` are
+  [DEPRECATED](/02-common-principles.html#definitions).
```
I started replacing internal relative links (e.g., `../02-common-principles.html#definitions`) with absolute versions (`/02-common-principles.html#definitions`). How does everyone feel about this change? I think it would be good to apply throughout the whole specification.

sappelhoff replied:
Why are you doing that? I thought it's an advantage to point to the relative files and have Mkdocs generate the suitable links for us? Pointing to html directly feels error prone but I can't put my finger on it right now :thinking:

tsalo replied:
One reason is that relative links depend on where the metadata field is rendered, so what might work for one file might not work for another. It should also make it easier to resolve links in external applications (e.g., NIDM-Terms).

sappelhoff replied:
I see :thinking: I couldn't find info on whether that's good or bad: https://www.mkdocs.org/user-guide/writing-your-docs/#internal-links

it's not really mentioned in the docs over there. So maybe we should carefully try it, and if it doesn't break anything it's fine.

tsalo commented on src/04-modality-specific-files/09-positron-emission-tomography.md:
```diff
@@ -130,18 +130,60 @@ which we divide into several categories:
+      "InstitutionName": "RECOMMENDED",
+      "InstitutionAddress": "RECOMMENDED",
+      "InstitutionalDepartmentName": "RECOMMENDED",
+      "BodyPart": "RECOMMENDED",
+   }
+) }}
+
 | **Key name**              | **Requirement level** | **Data type** | **Description**                                                                                                                                                                        |
 | ------------------------- | --------------------- | ------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
 | Manufacturer              | REQUIRED              | [string][]    | Scanner manufacturer (for example, `"Siemens"`).                                                                                                                                       |
 | ManufacturersModelName    | REQUIRED              | [string][]    | PET scanner model name (for example, `"mMR Biograph"`).                                                                                                                                |
 | Unit                      | REQUIRED              | [string][]    | Unit of the image file; please see BIDS main spec section 6. SI unit for radioactivity (Becquerel) should be used (for example, `"Bq/mL"`). Corresponds to DICOM Tag 0054, 1001 Units. |
 | InstitutionName           | RECOMMENDED           | [string][]    | The name of the institution in charge of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 0080 InstitutionName.                                     |
 | InstitutionAddress        | RECOMMENDED           | [string][]    | The address of the institution in charge of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 0081 InstitutionAddress.                               |
-| InstitutionDepartmentName | RECOMMENDED           | [string][]    | The department in the institution in charge of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 1040 Institutional Department Name.                 |
+| InstitutionalDepartmentName | RECOMMENDED           | [string][]    | The department in the institution in charge of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 1040 Institutional Department Name.                 |
```
I _think_ this was a typo in the PET BEP?

sappelhoff replied:
yes, most likely. fix: https://github.com/bids-standard/bids-specification/pull/771

tsalo commented on src/04-modality-specific-files/03-electroencephalography.md:
```diff
@@ -354,12 +405,30 @@ head.
 
+{{ MACROS___make_metadata_table(
+   {
+      "IntendedFor": (
+         "OPTIONAL",
+         "This identifies the MRI or CT scan associated with the electrodes, "
+         "landmarks, and fiducials.",
+      )
```
What does everyone think of this approach for extending descriptions for modality- or table-specific purposes?
